upup/pkg/kutil/rollingupdate_cluster_validate_drain.go

```diff
@@ -0,0 +1,313 @@
+				resultsMutex.Unlock()
+			}
+
+			defer wg.Done()
+
+			for k, group := range masterGroups {
+				rollingUpdateData := c.CreateRollingUpdateData(instanceGroups, false)
+
+				err := group.RollingUpdateDrainValidate(rollingUpdateData)
+				resultsMutex.Lock()
+				results[k] = err
+				resultsMutex.Unlock()
+
+				// TODO: Bail on error?
+			}
+		}()
+
+		wg.Wait()
+	}
+
+	// Upgrade nodes, with greater parallelism
+	{
+		var wg sync.WaitGroup
+
+		for k, nodeGroup := range nodeGroups {
+			wg.Add(1)
+			go func(k string, group *CloudInstanceGroup) {
+				resultsMutex.Lock()
+				results[k] = fmt.Errorf("function panic")
+				resultsMutex.Unlock()
+
+				defer wg.Done()
+
+				rollingUpdateData := c.CreateRollingUpdateData(instanceGroups, false)
+
+				err := group.RollingUpdateDrainValidate(rollingUpdateData)
+
+				resultsMutex.Lock()
+				results[k] = err
+				resultsMutex.Unlock()
+			}(k, nodeGroup)
+		}
+
+		wg.Wait()
+	}
+
+	for _, err := range results {
+		if err != nil {
+			return err
+		}
+	}
+
+	glog.Infof("\nRolling update completed!\n")
+	return nil
+}
+
+// CreateRollingUpdateData creates a RollingUpdateClusterDrainValidate struct.
+func (c *RollingUpdateClusterDrainValidate) CreateRollingUpdateData(instanceGroups *api.InstanceGroupList, isBastion bool) *RollingUpdateDataDrainValidate {
+	return &RollingUpdateDataDrainValidate{
+		Cloud:             c.Cloud,
+		Force:             c.Force,
+		Interval:          c.NodeInterval,
+		InstanceGroupList: instanceGroups,
+		IsBastion:         isBastion,
+		K8sClient:         c.K8sClient,
+		FailOnValidate:    c.FailOnValidate,
+		FailOnDrainError:  c.FailOnDrainError,
+		CloudOnly:         c.CloudOnly,
+		ClusterName:       c.ClusterName,
+	}
+}
+
+// RollingUpdateDrainValidate performs a rolling update on a list of ec2 instances.
+func (n *CloudInstanceGroup) RollingUpdateDrainValidate(rollingUpdateData *RollingUpdateDataDrainValidate) error {
+
+	// we should not get here, but hey I am going to check
+	if rollingUpdateData == nil {
+		return fmt.Errorf("rollingUpdate cannot be nil")
+	}
+
+	// Do not need a k8s client if you are doing cloud only
+	if rollingUpdateData.K8sClient == nil && !rollingUpdateData.CloudOnly {
+		return fmt.Errorf("rollingUpdate is missing a k8s client")
+	}
+
+	if rollingUpdateData.InstanceGroupList == nil {
+		return fmt.Errorf("rollingUpdate is missing a the InstanceGroupList")
+	}
+
+	c := rollingUpdateData.Cloud.(awsup.AWSCloud)
+
+	update := n.NeedUpdate
+	if rollingUpdateData.Force {
+		update = append(update, n.Ready...)
+	}
+
+	if !rollingUpdateData.IsBastion && rollingUpdateData.FailOnValidate && !rollingUpdateData.CloudOnly {
+		_, err := validate.ValidateCluster(rollingUpdateData.ClusterName, rollingUpdateData.InstanceGroupList, rollingUpdateData.K8sClient)
+		if err != nil {
+			return fmt.Errorf("cluster %s does not pass validation", rollingUpdateData.ClusterName)
+		}
+	}
+
+	for _, u := range update {
+
+		if !rollingUpdateData.IsBastion {
```

Comment:
Most trivially it'll avoid the nested if blocks.  It also keeps the function more readable IMO.